perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java

```diff
@@ -3923,42 +3999,106 @@ private void updateExistingGroupsWhileSynchronization(PerunSession sess, Group b
-					newParentGroup = subGroup;
-					break;
+		//check situation, where group is already under base group and the destination parent group is now base group too
+		if(oldParentGroupLogin == null && newParentGroup.equals(baseGroup)) {
+			return null;
+		}
+
+		return newParentGroup;
+	}
+
+	/**
+	 * Return parent group and its login for the candidate group in the group structure.
+	 *
+	 * @param sess perun session
+	 * @param candidateGroup candidate group
+	 * @param baseGroup base group in structure
+	 * @param loginAttributeName attribute name of login in the structure
+	 * @return new parent group of candidate group, if there is none, return base group
+	 */
+	private Group getNewParentGroupWithLogin(PerunSession sess, CandidateGroup candidateGroup, Group baseGroup, String loginAttributeName) {
+		Group newParentGroup = baseGroup;
+
+		//if parent group login is not null, we need to find this new parent group (otherwise it is base group so return empty pair)
+		if(candidateGroup.getParentGroupLogin() != null) {
+			try {
+				//we need to have vo to filter groups, base group has to be from the same vo as new parent of candidate group
+				Vo baseGroupVo = getPerunBl().getVosManagerBl().getVoById(sess, baseGroup.getVoId());
+				List<Group> groupsWithLogin = getPerunBl().getSearcherBl().getGroups(sess, baseGroupVo, Collections.singletonMap(loginAttributeName, candidateGroup.getParentGroupLogin()));
+				//if there is missing the destination parent group, set base group as parent, there could be missing parent group in the extSource data
+				if (!groupsWithLogin.isEmpty()) {
+					//if there are more than one group with login, there is no way how to choose correct behavior
+					if (groupsWithLogin.size() > 1)
+						throw new InternalErrorException("More than 1 group has login attribute '" + loginAttributeName + "' set with the same login " + candidateGroup.getParentGroupLogin());
+					//our new parent group is exactly the one with the login set
+					return groupsWithLogin.get(0);
 				}
+			} catch (VoNotExistsException ex) {
+				throw new InternalErrorException("Vo of base group: " + baseGroup + " can't be found. Unexpected situation, can't continue.");
+			} catch (AttributeNotExistsException ex) {
+				throw new InternalErrorException("Can't find attribute definition for '" + loginAttributeName + "'.");
 			}
 		}
+
 		return newParentGroup;
 	}
 
+	/**
+	 * Return parent group and its login for the group to update in the group structure.
+	 *
+	 * @param sess perun session
+	 * @param groupToUpdate group to update
+	 * @param baseGroup base group in structure
+	 * @param loginAttributeName attribute name of login in the structure
+	 * @return current parent group login of group to update, null if there is none (it means base group should be parent)
+	 */
+	private String getOldParentGroupWithLogin(PerunSession sess, Group groupToUpdate, Group baseGroup, String  loginAttributeName) {
```

Comment:
This name and Javadoc are misleading. You are returning only the parent group login not the parent group with its login.

Reply:
Name and javadoc of this method and also the method getNewParentGroupWithLogin was changed.